abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+  - `i32 (uint32_t*) return_cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get shared data identified by a key (`key_data`, `key_size`). The compare-and-switch value
+(`return_cas`) is returned and can be used when updating the value with `proxy_set_shared_data`.
+
+
+### `proxy_set_shared_data`
+
+* params:
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char*) value_data`
+  - `i32 (size_t) value_size`
+  - `i32 (uint32_t) cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set shared data identified by a key (`key_data`, `key_size`) to a value (`value_data`,
+`value_size`). If compare-and-switch value (`cas`) is set, then it must match the current value in
+order to update to succeed.
+
+
+## Shared queue
+
+### `proxy_register_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Register a shared queue using a given name (`queue_name_data`, `queue_name_size`). It can be
+referred to in `proxy_enqueue_shared_queue` and `proxy_dequeue_shared_queue` using returned unique
+queue identifier (`queue_id`).
+
+
+### `proxy_resolve_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resolves existing shared queue using a given name (`queue_name_data`, `queue_name_size`). It can be
+referred to in `proxy_enqueue_shared_queue` and `proxy_dequeue_shared_queue` using returned unique
+queue identifier (`queue_id`).
+
+
+### `proxy_dequeue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const char**) payload_data`
+  - `i32 (size_t*) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get data (`payload_data`, `payload_size`) to the end of the queue (`queue_id`).
+
+
+### `proxy_enqueue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const char*) payload_data`
+  - `i32 (size_t) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Add data (`payload_data`, `payload_size`) to the front of the queue (`queue_id`).
+
+
+### `proxy_remove_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Remove queue (`queue_id`).
+
+
+## Stats/metrics
+
+### `proxy_define_metric`
+
+* params:
+  - `i32 (proxy_metric_type_t) metric_type`
+  - `i32 (const char*) metric_name_data`
+  - `i32 (size_t) metric_name_size`
+  - `i32 (uint32_t*) return_metric_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Define a metric using a given name (`metric_name_data`, `metic_name_size`). It can be referred to in
+`proxy_get_metric`, `proxy_increment_metric` and `proxy_record_metric` using returned unique metric
+identifier (`metric_id`).
+
+
+### `proxy_get_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
```

Comment:
In order to read histograms we can add an i32 "value" field which would be the histogram value to read the count from.

Reply:
The problem is histograms are implementation specific in terms of whether they are HDR, fixed bucket, support summary percentile only, etc. I suspect that for histograms we are going to need a facility to actually read out the buckets, percentiles, etc. I don't see any other sensible way to do this that would support different host types. With that said, I think it's concievable that you can jam other metric types into this API if you squint really hard, but I just don't see how we won't wind up with some histogram specific ABI eventually.